### ISSUE TITLE
Precision bug in `optax.softmax_cross_entropy_with_integer_labels`

### ISSUE THREAD

**Issue by MasterSkepticista** [google-deepmind/optax#1020]
Loss values are quantized when `logits` are supplied as `bfloat16` due to `log_normalizers` getting quantized after computation.
https://github.com/google-deepmind/optax/blob/6f1291108f1ec2af3ce1f0191296d5b2a56cefab/optax/losses/_classification.py#L265-L271

Example:
```python
b, c = 8, 100
key = jax.random.key(42)
loss_fn = optax.softmax_cross_entropy_with_integer_labels

# bug
logits_b16 = jax.random.uniform(key, (b, c), jnp.bfloat16)
labels = jax.random.randint(key, (b,), 0, c)

print(loss_fn(logits_b16, labels))
# [4.75 5 4.59375 5.0625 4.4375 4.15625 4.3125 4.53125]

print(loss_fn(logits_b16.astype(jnp.float32), labels))
# [4.7477226 4.9903674 4.592032  5.0644608 4.4306226 4.1690445 4.301924 4.52517  ]
```

**Possible fix**: cast `logits` to `float32` within the function.
There is a performance penalty for large values (but practically negligible?). No penalty if `logits` are already `float32`.

```python
b, c = 4096, 50_000  # ~200M elements
key = jax.random.key(42)
jitted_fn = jax.jit(optax.softmax_cross_entropy_with_integer_labels)

labels = jax.random.randint(key, (b,), 0, c)
logits_b16 = jax.random.uniform(key, (b, c), jnp.bfloat16)

# Warmup.
jitted_fn(logits_b16, labels).block_until_ready()

%timeit -n 100 jitted_fn(logits_b16, labels).block_until_ready()
1.17 ms ± 2.98 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit -n 100 jitted_fn(logits_b16.astype(jnp.float32), labels).block_until_ready()
4.06 ms ± 1.79 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by fabianp** [comment]
hey @MasterSkepticista . Perhaps I'm missing something, but it seems to me that the difference between the result in bfloat and float32 is what you would expect from a low precision type like bfloat. In your example, the relative error is around 1e-3, which is what I would expect since bfloat16 has only 3 significant decimal digits.

<img width="383" alt="image" src="https://github.com/user-attachments/assets/f426fe8f-a2c5-49cf-b2fe-752290e008bb">


I'm closing this for now but feel free to reopen.
